db/view: Require rf_rack_valid_keyspaces when creating materialized view

dawmd opened the pull request:
Materialized views are currently in the experimental phase and using them
in tablet-based keyspaces requires starting Scylla with an experimental feature,
`views-with-tablets`. Any attempts to create a materialized view or secondary
index when it's not enabled will fail with an appropriate error.

After considerable effort, we're drawing close to bringing views out of the
experimental phase, and the experimental feature will no longer be needed.
However, materialized views in tablet-based keyspaces will still be restricted,
and creating them will only be possible after enabling the configuration option
`rf_rack_valid_keyspaces`. That's what we do in this PR.

In this patch, we adjust existing tests in the tree to work with the new
restriction. That shouldn't have been necessary because we've already seemingly
adjusted all of them to work with the configuration option, but some tests hid
well. We fix that mistake now.

After that, we introduce the new restriction. What's more, when starting Scylla,
we verify that there is no materialized view that would violate the contract.
If there are some that do, we list them, notify the user, and refuse to start.

High-level implementation strategy:

1. Name the restrictions in form of a function.
2. Adjust existing tests.
3. Restrict materialized views by both the experimental feature
   and the configuration option. Add validation test.
4. Drop the requirement for the experimental feature. Adjust the added test
   and add a new one.
5. Update the user documentation.

Fixes scylladb/scylladb#23030

Backport: 2025.4, as we are aiming to support materialized views for tablets from that version.

calebxyz commented:
@piodul can we start reviewing this one ?

piodul commented:
@dawmd We won't need this PR if we merge https://github.com/scylladb/scylladb/pull/23757, am I correct?

If so, then please close this PR.

(IIRC we planned to do it in two steps: make enablement conditional and then remove the experimental feature, but the other PR does both)

dawmd commented:
> @dawmd We won't need this PR if we merge #23757, am I correct?
> 
> If so, then please close this PR.
> 
> (IIRC we planned to do it in two steps: make enablement conditional and then remove the experimental feature, but the other PR does both)

I updated this PR instead. I think it has a clearer structure and is easier to review. I moved the necessary changes from the other PR here.

For what it's worth, I removed one of the tests. Its purpose is no longer clear to me and I don't think it brings much value, while taking significant time to execute.

piodul commented on db/view/view.cc:
```diff
@@ -3689,27 +3689,15 @@ sstring build_status_to_sstring(build_status status) {
 }
 
 void validate_view_keyspace(const data_dictionary::database& db, std::string_view keyspace_name) {
-    const bool tablet_views_enabled = db.features().views_with_tablets;
```
This is wrong. We want to remove the experimental flag but we want to keep the cluster feature, and keep checking the cluster feature.

Perhaps we should add a clear message - if the cluster feature is not enabled yet, an appropriate message mentioning it should be printed.

dawmd replied:
Let's add the message in a separate PR. I'm not sure if it's going to be a subject of discussion or not, but it's a detail, so it might be a better approach.

calebxyz replied:
Agree about the message to user, @piodul after this change were good to queue ?

piodul commented on db/view/view.cc:
```diff
@@ -3687,5 +3688,21 @@ sstring build_status_to_sstring(build_status status) {
+
+    const bool uses_tablets = db.find_keyspace(keyspace_name).get_replication_strategy().uses_tablets();
+
+    if (!rf_rack_valid_keyspaces && uses_tablets) {
+        throw std::logic_error("Materialized views and secondary indexes are not supported on base tables with tablets. "
+                "To be able to use them, enable the configuration option `rf_rack_valid_keyspaces`.");
+    }
```
There is also a warning in `create_keyspace_statement.cc`:

```c++
        if (rs->uses_tablets()) {
            warnings.push_back(
                "Tables in this keyspace will be replicated using Tablets "
                "and will not support Materialized Views, Secondary Indexes and counters features. "
                "To use Materialized Views, Secondary Indexes or counters, drop this keyspace and re-create it "
                "without tablets by adding AND TABLETS = {'enabled': false} to the CREATE KEYSPACE statement.");
            if (ksm->initial_tablets().value()) {
                warnings.push_back("Keyspace `initial` tablets option is deprecated.  Use per-table tablet options instead.");
            }
        }
```

I suggest adjusting it appropriately.

dawmd commented:
@scylladbbot trigger-ci

dawmd commented:
Version 5.0:
* Adjusted new materialized views that explicitly disabled enforcing RF-rack-valid keyspaces.
* Marked one of those tests as skipped because it requires more thought to adjust it, and I haven't been able to do that yet.

dawmd commented:
> #### 🔴 CI State: FAILURE
> Mode 	Stage 	Status 	Node
> 	[Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/20623/) 	✅ 	spider9.cloudius-systems.com
> dev 	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/20623/) 	✅ 	spider12.cloudius-systems.com
> 	[test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/20623/) 	✅ 	spider12.cloudius-systems.com
> 	[Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/20623/)
> 🔹 test/boost/schema_loader_test.cc
> 🔹 cluster/mv/test_mv_simple
> 🔹 cluster/mv/test_mv_topology_change
> 🔹 cluster/test_view_building_coordinator 	✅ 	spider12.cloudius-systems.com
> release 	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/20623/) 	✅ 	spider1.cloudius-systems.com
> 	[test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/20623/) 	❌ 	spider1.cloudius-systems.com
> debug 	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/20623/) 	✅ 	spider5.cloudius-systems.com
> 	[test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/20623/) 	❌ 	spider5.cloudius-systems.com
> dtest 	[dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/19360/) 	✅ 	
> 	[dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/11216/) 	✅ 	
> 	[dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/7917/) 	✅ 	
> #### Failed Tests (2/78055):
> 
>     * [test_keep_alive[False].release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20623/testReport/junit/alternator/test_manual_requests/release___test_py___release___test_keep_alive_False__release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_keep_alive%5BFalse%5D.release.2)
> 
>     * [test_random_failures[stop_after_starting_cdc_generation_service-add_index].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/20623/testReport/junit/cluster.random_failures/test_random_failures/debug___test_py___debug___test_random_failures_stop_after_starting_cdc_generation_service_add_index__debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_random_failures%5Bstop_after_starting_cdc_generation_service-add_index%5D.debug.2)
> 
> 
> #### Build Details:
> 
>     * Duration: 5 hr 15 min
> 
> 
> **Note:** To re-trigger CI for this PR, comment: `@scylladbbot trigger-ci`

The alternator test is scylladb/scylladb#26298. The other failure is valid and wasn't detected before because events in the test are randomised.

dawmd commented:
Also, I'm aware that disabling/skipping tests or test cases isn't a good idea. I went through this approach because this PR is still pending:
* If we decide that it must be merged immediately, CI will pass and we won't be dealing with all kinds of problems.
* If skipping tests or test cases is a no-go, we need to spend a bit more time on it anyway.

Hence, this seems to be the most reasonable solution to me at the moment.

pehala commented on test/cluster/mv/test_mv_simple.py:
```diff
@@ -0,0 +1,136 @@
+
+    # Views no longer depend on the experimental feature `views-with-tablets`,
+    # but let's keep these test cases to make sure it's really not needed anymore.
+    await run_test(False, False)
+    await run_test(False, True)
+    await run_test(True, False)
+    await run_test(True, True)
```
Why is this not a parametrization?

pehala commented on test/cluster/mv/test_mv_simple.py:
```diff
@@ -0,0 +1,136 @@
+                        "AND tablets = {'enabled': true}")
+    await cql.run_async("CREATE TABLE ks.t (p int PRIMARY KEY, v int)")
+
+    if view_type == "view":
+        await cql.run_async("CREATE MATERIALIZED VIEW ks.mv "
+                            "AS SELECT * FROM ks.t "
+                            "WHERE p IS NOT NULL AND v IS NOT NULL "
+                            "PRIMARY KEY (v, p)")
+    elif view_type == "index":
+        await cql.run_async("CREATE INDEX i ON ks.t(v)")
+    else:
+        logger.error(f"Unexpected view type: {view_type}")
+        assert False
```
This seems to be shared among multiple tests, including the parametrization, maybe it would be worth extracting it and having this be a parametrised fixture

dawmd replied:
I'm open to this, but since it's more of a stylistic thing, let's do that in a follow-up. :) Thanks for the suggestion, Petr!

pehala commented on test/cluster/mv/test_mv_simple.py:
```diff
@@ -0,0 +1,136 @@
+        logger.debug("Obtained CassandraSession object")
+
+        # We just want to validate the statements. We don't need to wait.
+        assert hasattr(cql.cluster, "max_schema_agreement_wait")
```
Can this ever fail?

dawmd replied:
I'm not sure. Some tests access it blindly, but some do not. It's more of a personal choice here, though I guess it doesn't make much difference (if the attribute doesn't exist, trying to access it should produce a clear error anyway AFAIK).

pehala reviewed (commented):
Small nits for the testpy tests structure

Jadw1 commented on test/cluster/mv/test_mv_topology_change.py:
```diff
@@ -256,6 +256,11 @@ async def do_writes() -> int:
 @pytest.mark.asyncio
 @pytest.mark.parametrize("delayed_replica", ["base", "mv"])
 @pytest.mark.parametrize("altered_dc", ["dc1", "dc2"])
+# FIXME: The test relies on cross-rack tablet migrations. They're forbidden when the configuration option
```
Is RF change currently supported with `rf_rack_valid_keyspaces` on? If no there should be an issue for it, which should be mentioned here (you can also use xfail mark).

Jadw1 replied:
Is RF change currently supported with `rf_rack_valid_keyspaces` on? If no there should be an issue for it, which should be mentioned here (you can also use xfail mark).

piodul commented on replica/database.hh:
```diff
@@ -2090,6 +2090,23 @@ public:
     //   must contain a complete list of racks and data centers in the cluster.
     void check_rf_rack_validity(const bool enforce_rf_rack_valid_keyspaces, const locator::token_metadata_ptr) const;
+
+    /// Verify that all existing materialized views are valid.
```
The comment needs to be updated, it says that the function throws an exception but in fact it only prints a warning. Let's fix it in a follow-up.
